packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/ExtensibleEnumSerializationProvider.cs

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using Microsoft.Generator.CSharp.Expressions;
+using Microsoft.Generator.CSharp.Input;
+using Microsoft.Generator.CSharp.Primitives;
+using Microsoft.Generator.CSharp.Providers;
+using static Microsoft.Generator.CSharp.Snippets.Snippet;
+
+namespace Microsoft.Generator.CSharp.ClientModel.Providers
+{
+    /// <summary>
+    /// This defines a class with extension methods for enums to convert an enum to its underlying value, or from its underlying value to an instance of the enum
+    /// </summary>
+    internal partial class ExtensibleEnumSerializationProvider : TypeProvider
+    {
+        private readonly InputEnumType _enumType;
+        private TypeProvider? _enumProvider;
+        private TypeProvider EnumProvider => _enumProvider ??= ClientModelPlugin.Instance.TypeFactory.CreateEnum(_enumType);
+
+        protected override string GetNamespace() => ClientModelPlugin.Instance.Configuration.ModelNamespace;
+
+        public ExtensibleEnumSerializationProvider(InputEnumType enumType)
+        {
+            Debug.Assert(enumType.IsExtensible);
+            _enumType = enumType;
+        }
+
+        protected override string BuildRelativeFilePath()
+        {
+            return Path.Combine("src", "Generated", "Models", $"{EnumProvider.Name}.Serialization.cs");
+        }
+
+        protected override string BuildName() => EnumProvider.Name;
+
+        public ValueExpression ToSerial(ValueExpression enumExpression)
+        {
+            var serialMethodName = EnumProvider.EnumUnderlyingType.Equals(typeof(string)) ? nameof(object.ToString) : $"ToSerial{EnumProvider.EnumUnderlyingType.Name}";
+            return enumExpression.Invoke(serialMethodName);
+        }
+
+        public ValueExpression ToEnum(ValueExpression valueExpression)
+            => New.Instance(Type, valueExpression);
+
+        protected override TypeSignatureModifiers GetDeclarationModifiers() => EnumProvider.DeclarationModifiers;
+
+        protected override MethodProvider[] BuildMethods()
+        {
+            // for string-based extensible enums, we are using `ToString` as its serialization
+            // for non-string-based extensible enums, we need a method to serialize them
+            if (!EnumProvider.EnumUnderlyingType.Equals(typeof(string)))
```

Comment:
So would it make sense if always put the `override ToString()` method in the serialization partial type?
It is a "serialization" for string

Reply:
ToString belongs to the typeprovider the fact that we use it for serialization is just by chance.